core/src/main/java/org/infinispan/remoting/inboundhandler/NonTotalOrderPerCacheInboundInvocationHandler.java

```diff
@@ -27,15 +57,44 @@ public void handle(CacheRpcCommand command, Reply reply, DeliverOrder order) {
       }
       try {
          final int commandTopologyId = extractCommandTopologyId(command);
-         final boolean waitTransactionalData = command.getCommandId() != StateRequestCommand.COMMAND_ID;
          final boolean onExecutorService = executeOnExecutorService(order, command);
-         handleRunnable(createDefaultRunnable(command, reply, commandTopologyId, waitTransactionalData, onExecutorService), onExecutorService);
+         final BlockingRunnable runnable;
+
+         switch (command.getCommandId()) {
+            case SingleRpcCommand.COMMAND_ID:
+               //lock ownership needs the first topology. we need to check it before invoking the getLockPromise()
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+                  reply.reply(CacheNotFoundResponse.INSTANCE);
+                  return;
+               }
+               runnable = createLockAwareRunnable(command, reply, commandTopologyId, true, onExecutorService,
+                                                  getLockPromise((SingleRpcCommand) command));
+               break;
+            case MultipleRpcCommand.COMMAND_ID:
+               //lock ownership needs the first topology. we need to check it before invoking the getLockPromise()
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+                  reply.reply(CacheNotFoundResponse.INSTANCE);
+                  return;
+               }
+               runnable = createLockAwareRunnable(command, reply, commandTopologyId, true, onExecutorService,
+                                                  getLockPromise((MultipleRpcCommand) command));
```

Comment:
I'm not sure about this, but I think it may be better to create a separate runnable for each command or it would introduce unnecessary delays. OTOH, since MultipleRpcCommands are only used by the replication queue, which has huge latencies anyway, it probably doesn't matter.